### PR TITLE
Adding support for reading ranklib data for ranking models @rev lastmansleeping@

### PR DESCRIPTION
**Pull request by mohazahran**
Adding support for ranklib data format in ml4ir

Previous PR -> https://github.com/salesforce/ml4ir/pull/53

### CODE REVIEW

File: python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        if feature_config.endswith(".yaml"):
+            feature_config = file_io.read_yaml(feature_config)
+        else:
+            feature_config = yaml.safe_load(feature_config)
+
+        return ExampleFeatureConfig(feature_config)
+
+    def test_ranklib_in_ml4ir(self):
+        exFeatureConfig = self.parse_config(TFRecordTypeKey.EXAMPLE, self.feature_config_yaml)
+        IO = local_io.LocalIO()
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py:101]: [Minor] @mohazahran Typically, it is preferred to use lower case for objects.. You can use `file_io` or `local_io`

File: python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            preprocessing_keys_to_fns={},
+            logger=None,
+            keep_additional_info=KEEP_ADDITIONAL_INFO,
+            gl_2_clicks=GL_2_CLICKS,
+            non_zero_features_only=NON_ZERO_FEATURES_ONLY
+        )
+
+
+        chk = [e for e in dataset.train]
+        assert len(chk) == 156
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py:118]: @mohazahran Can you explicitly set the `max_sequence_size` argument?

File: python/ml4ir/base/config/parse_args.py [hunk 3]
```diff
@@ -302,6 +302,27 @@ def define_args(self):
             help="Whether to track model performance and hyperparameters.",
         )
 
+        self.add_argument(
+            "--gl_2_clicks",
```
Comment (lastmansleeping) [inline python/ml4ir/base/config/parse_args.py:306]: @mohazahran Thinking more about this, I feel it makes sense to add this to the feature_config yaml under the label feature. What do you think?

File: python/ml4ir/base/config/parse_args.py [hunk 4]
```diff
@@ -302,6 +302,27 @@ def define_args(self):
...
+            type=int,
+            default=1,
+            help="Convert graded relevance to clicks (only max relevant document is considered clicked) 1 to convert",
+        )
+
+        self.add_argument(
+            "--non_zero_features_only",
+            type=int,
+            default=0,
+            help="Only non zero features are stored. 1 for yes, 0 otherwise",
```
Comment (lastmansleeping) [inline python/ml4ir/base/config/parse_args.py:316]: @mohazahran Can you also specify that this is only for ranklib data format?

File: python/ml4ir/base/data/ranklib_to_ml4ir.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
+import pandas as pd
+import argparse
+import logging
+import numpy as np
+
+
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/ranklib_to_ml4ir.py:6]: @mohazahran Can we name this file better? Perhaps `ranklib_helper`?

File: python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+              shape: null
+            serving_info:
+              required: true
+              default_value: 0.0
+            tfrecord_type: sequence
+        '''
+
+    def parse_config(self, tfrecord_type: str, feature_config) -> ExampleFeatureConfig:
+        if feature_config.endswith(".yaml"):
+            feature_config = file_io.read_yaml(feature_config)
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py:93]: @mohazahran Why not save this to a file?

File: python/ml4ir/base/data/relevance_dataset.py [hunk 7]
```diff
@@ -28,6 +29,9 @@ def __init__(
         use_part_files: bool = False,
         parse_tfrecord: bool = True,
         logger: Optional[Logger] = None,
+        keep_additional_info: int = 0,
+        gl_2_clicks: int = 1,
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/relevance_dataset.py:33]: @mohazahran Let's see if we can add `gl_2_clicks` to the feature_config itself under label.

**Review by lastmansleeping — CHANGES_REQUESTED** [review]
Looking good @mohazahran 
One change I recommend is adding the `gl_2_clicks` argument to the feature_config yaml itself under `label`. Might be cleaner. Thoughts?

File: python/ml4ir/base/config/parse_args.py (hunk 3 above)
Comment (mohazahran) [reply python/ml4ir/base/config/parse_args.py:306]: That's possible. However, that requires supporting the conversion to clicks to other data format (CSV and TFrecords) which are not currently supporting this.

File: python/ml4ir/base/data/ranklib_to_ml4ir.py (hunk 5 above)
Comment (mohazahran) [reply python/ml4ir/base/data/ranklib_to_ml4ir.py:6]: Well, the initial purpose of this file is to be a stand alone script for converting ranklib to ml4ir. It still can serve this purpose as it can be called with arguments to convert to csv. With that in mind, its name is appropriate no?

File: python/ml4ir/applications/ranking/tests/data/ranklib_test_data/feature_config.yaml [hunk 8]
```diff
@@ -0,0 +1,76 @@
+query_key:
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/tests/data/ranklib_test_data/feature_config.yaml:1]: @mohazahran Recommend renaming the folder to `tests/data/ranklib` instead in line with `csv/` and `tfrecord/`.

Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/tests/data/ranklib_test_data/feature_config.yaml:1]: Also recommend moving this config to `tests/data/configs/...`

File: python/ml4ir/applications/ranking/tests/test_convert_label_to_clicks.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import tensorflow as tf
+import numpy as np
+import warnings
+warnings.filterwarnings("ignore")
+
+class TestClickConversion(unittest.TestCase):
+    def setUp(self):
+        pass
+
+    def test_1(self):
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/tests/test_convert_label_to_clicks.py:12]: @mohazahran You should rename this function to be more descriptive and add a pydoc comment explaining what it is testing.

The test looks good though. :)

File: python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+                break
+        assert non_one_hot == True
+        assert len(chk) == 49
+
+        exFeatureConfig = self.parse_config(TFRecordTypeKey.SEQUENCE_EXAMPLE, self.feature_config_yaml_convert_to_clicks, io)
+        preprocessing_keys_to_fns = {}
+        if exFeatureConfig.get_label()['preprocessing_info'][0]['fn'] == 'convert_label_to_clicks':
+            preprocessing_keys_to_fns['convert_label_to_clicks'] = convert_label_to_clicks
+
+        dataset = RelevanceDataset(
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py:89]: @mohazahran Break it up into smaller tests. I understand what you are trying to do, but will be better if you break it up IMO.

File: python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+            assert max(e[1][0]).numpy() == 1
+        assert len(chk) == 49
+
+        chk = [e for e in dataset.test]
+        for e in chk:
+            assert max(e[1][0]).numpy() == 1
+        assert len(chk) == 49
+
+
+
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py:118]: @mohazahran Extra new lines?

File: python/ml4ir/base/data/ranklib_reader.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    - selects relevant features
+    - creates Dataset X and y
+
+    Current execution plan:
+        1. Conver ranklib to a dataframe
+        2. Convert each query into tf.train.SequenceExample protobufs
+        3. Write the protobufs into a .tfrecord file
+        4. Load .tfrecord file into a TFRecordDataset and parse the protobufs
+
+    Args:
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/ranklib_reader.py:40]: @mohazahran Can you use the numpy doc format?
https://numpydoc.readthedocs.io/en/latest/format.html

File: python/ml4ir/base/data/ranklib_reader.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+        - data_dir: Path to directory containing csv files to read
+        - feature_config: ml4ir.config.features.FeatureConfig object extracted from the feature config
+        - tfrecord_dir: Path to directory where the serialized .tfrecord files will be stored
+        - batch_size: int value specifying the size of the batch
+        - use_part_files: bool value specifying whether to look for part files
+        - max_sequence_size: int value specifying max number of records per query
+        - logger: logging object
+        - keep_additional_info: Option to keep additional info (All info after the "#") 1 to keep, 0 to ignore
+        - non_zero_features_only: Only non zero features are stored. 1 for yes, 0 otherwise
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/ranklib_reader.py:49]: @mohazahran Can this not be done from the feature config YAML?

File: python/ml4ir/base/data/ranklib_to_ml4ir.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+#Constants for the column names given to the query id and graded relevance.
+QUERY_ID_NAME = 'qid'
+QUERY_TYPE = 'str'
+RELEVANCE_NAME = 'relevance'
+
+max_f_id = 0
+def process_line(line, keep_additional_info):
+    ''' Takes an input line in ranklib format and returns a row in ml4ir format.'''
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/ranklib_to_ml4ir.py:14]: @mohazahran Add pydocs here.

File: python/ml4ir/base/data/ranklib_to_ml4ir.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+            except:
+                val = fv.split(':')[1].strip()
+            r['f_' + feat] = val
+            global max_f_id
+            if int(feat) > max_f_id:
+                max_f_id = int(feat)
+    return r
+
+def convert(input_file, keep_additional_info, gl_2_clicks, non_zero_features_only, add_dummy_rank_column = True):
+    ''' Convert the input file with the specified parameters into ml4ir format.
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/ranklib_to_ml4ir.py:42]: @mohazahran pydocs here as well.

File: python/ml4ir/base/data/ranklib_to_ml4ir.py (hunk 5 above)
Comment (lastmansleeping) [reply python/ml4ir/base/data/ranklib_to_ml4ir.py:6]: Can we call it `ranklib_to_tfrecord` perhaps?

File: python/ml4ir/base/data/ranklib_to_ml4ir.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        df = pd.DataFrame(rows, columns=columns)
+        df.replace(np.nan, 0, inplace=True)
+    else:
+        df = pd.DataFrame(rows)
+    if int(gl_2_clicks) == 1:
+        groups = df.groupby(QUERY_ID_NAME)
+        for gname, group in groups:
+            df.loc[df[QUERY_ID_NAME] == gname, RELEVANCE_NAME] = (
+                    df.loc[df[QUERY_ID_NAME] == gname].relevance == max(group.relevance)).astype(int)
+    # ml4ir expects a pre-ranking. Adding a dummy pre-ranking to match format.
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/ranklib_to_ml4ir.py:63]: Would prepend `NOTE: ` here.

File: python/ml4ir/base/data/ranklib_to_ml4ir.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+
+if __name__ == "__main__":
+    # Setup logging
+    logger = logging.getLogger()
+    logger.setLevel(logging.DEBUG)
+    logging.debug("Logger is initialized...")
+
+    # parsing arguments
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--input_file', type=str, default='ml4ir/applications/ranking/tests/data/train/sample.txt', help='ranklib input file path')
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/ranklib_to_ml4ir.py:86]: @mohazahran Would be useful to convert a directory of ranklib files, no?

File: python/ml4ir/base/data/ranklib_to_ml4ir.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+    for line in f:
+        rows.append(process_line(line, keep_additional_info))
+    f.close()
+    if non_zero_features_only:
+        columns = [QUERY_ID_NAME, RELEVANCE_NAME] + ['f_' + str(i) for i in range(max_f_id)]
+        df = pd.DataFrame(rows, columns=columns)
+        df.replace(np.nan, 0, inplace=True)
+    else:
+        df = pd.DataFrame(rows)
+    if int(gl_2_clicks) == 1:
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/ranklib_to_ml4ir.py:58]: @mohazahran [Minor] Why not just use booleans here?

File: python/ml4ir/base/data/ranklib_to_ml4ir.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
+import pandas as pd
+import argparse
+import logging
+import numpy as np
+
+
+#Constants for the column names given to the query id and graded relevance.
+QUERY_ID_NAME = 'qid'
+QUERY_TYPE = 'str'
+RELEVANCE_NAME = 'relevance'
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/ranklib_to_ml4ir.py:10]: @mohazahran You should accept all these 3 constants as command line arguments with these values as defaults. That is the best middle ground.

File: python/ml4ir/base/pipeline.py [hunk 20]
```diff
@@ -236,6 +212,11 @@ def get_relevance_dataset(self, preprocessing_keys_to_fns={}) -> RelevanceDatase
 
         NOTE: Override this method to create custom dataset objects
         """
+
+        if 'preprocessing_info' in self.feature_config.get_label():
```
Comment (lastmansleeping) [inline python/ml4ir/base/pipeline.py:216]: You should move this to `ml4ir/applications/ranking/pipeline.py`. You can use `super().get_relevance_dataset(...)` method to do this.

File: python/ml4ir/base/data/tfrecord_reader.py [hunk 21]
```diff
@@ -258,7 +258,9 @@ def crop_fn():
 
         if not required_fields_only:
             # Check if label is one-hot and correctly masked
-            tf.debugging.assert_equal(tf.cast(tf.reduce_sum(labels), tf.float32), tf.constant(1.0))
+            #tf.debugging.assert_equal(tf.cast(tf.reduce_sum(labels), tf.float32), tf.constant(1.0))
+            #commeting the previous line as it filters out all queries with non 1-hot label vectors.
+            pass
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/tfrecord_reader.py:263]: @mohazahran I recommend deleting this whole if block actually.

File: python/ml4ir/base/data/tfrecord_reader.py [hunk 22]
```diff
@@ -344,7 +346,9 @@ def read(
 
     # Parse the protobuf data to create a TFRecordDataset
     dataset = data.TFRecordDataset(tfrecord_files)
+
     if parse_tfrecord:
+        #dataset = dataset.map(parse_fn)
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/tfrecord_reader.py:351]: @mohazahran Remove this comment?

**Review by lastmansleeping — CHANGES_REQUESTED** [review]
Awesome job @mohazahran 
Some final changes:
- Instead of a script on `ranklib_to_ml4ir`, make `ranklib_reader` a script from command line as it converts data to TFRecords and also uses FeatureConfig.
- Consider how we can remove the two additional arguments by using the FeatureConfig perhaps
- Add pydocs for all methods in the numpy format(https://numpydoc.readthedocs.io/en/latest/format.html)
- Move labels to clicks test to `ml4ir/applications/ranking/tests/test_preprocessing.py`

File: python/ml4ir/base/data/ranklib_reader.py (hunk 13 above)
Comment (mohazahran) [reply python/ml4ir/base/data/ranklib_reader.py:49]: These are passed as command line arguments

File: python/ml4ir/base/data/ranklib_to_ml4ir.py (hunk 17 above)
Comment (mohazahran) [reply python/ml4ir/base/data/ranklib_to_ml4ir.py:86]: Will add a function to convert directories as well.

File: python/ml4ir/base/data/ranklib_to_ml4ir.py (hunk 19 above)
Comment (mohazahran) [reply python/ml4ir/base/data/ranklib_to_ml4ir.py:10]: I can read the names from the feature_config. What do you think?

File: python/ml4ir/base/pipeline.py (hunk 20 above)
Comment (mohazahran) [reply python/ml4ir/base/pipeline.py:216]: I'm not sure that I follow. I added these lines to get_relevance_dataset(...)

**Comment by lastmansleeping** [comment]
@mohazahran Please merge with master

File: python/ml4ir/base/pipeline.py (hunk 20 above)
Comment (mohazahran) [reply python/ml4ir/base/pipeline.py:216]: @lastmansleeping can you elaborate a little bit more.

File: python/ml4ir/applications/ranking/tests/test_feature_processing.py [hunk 23]
```diff
@@ -93,3 +93,25 @@ def test_split_and_pad_string(self):
         expected_split_text = tf.constant(["", "A", "BC", "ab", "c1", "23!@", "#"] + [""]*18)
 
         assert tf.reduce_all(tf.equal(split_text, expected_split_text))
+
+    def testing_click_conversion(self):
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/tests/test_feature_processing.py:97]: @mohazahran Please add pydocs for the test.

File: python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+from ml4ir.base.io import file_io, local_io
+
+warnings.filterwarnings("ignore")
+
+INPUT_DIR = "ml4ir/applications/ranking/tests/data/ranklib_test_data/"
+KEEP_ADDITIONAL_INFO = 0
+NON_ZERO_FEATURES_ONLY = 0
+
+
+class TestRanklibConversion(unittest.TestCase):
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/tests/test_ranklib_in_ml4ir.py:21]: @mohazahran Please add pydocs for the test.

File: python/ml4ir/applications/ranking/tests/test_ranklib_to_ml4ir.py [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+import unittest
+import os
+import warnings
+from ml4ir.base.data import ranklib_helper
+import pandas as pd
+
+
+warnings.filterwarnings("ignore")
+
+INPUT_FILE = "ml4ir/applications/ranking/tests/data/ranklib_test_data/train/sample.txt"
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/tests/test_ranklib_to_ml4ir.py:11]: @mohazahran Recommend moving this test to the other ranklib one itself.

File: python/ml4ir/base/data/ranklib_reader.py [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+
+    Current execution plan:
+        1. Conver ranklib to a dataframe
+        2. Convert each query into tf.train.SequenceExample protobufs
+        3. Write the protobufs into a .tfrecord file
+        4. Load .tfrecord file into a TFRecordDataset and parse the protobufs
+
+        Parameters
+        ----------
+            data_dir: str
```
Comment (lastmansleeping) [inline python/ml4ir/base/data/ranklib_reader.py:42]: @mohazahran I think you need to align the variable names with `Param...`. I see an extra tab. More of a nitpick. The docs might still work. Not sure.

File: python/ml4ir/base/features/preprocessing.py [hunk 27]
```diff
@@ -18,7 +18,7 @@ def __init__(self):
         self.key_to_fn = {
             preprocess_text.__name__: preprocess_text,
             split_and_pad_string.__name__: split_and_pad_string,
-            natural_log.__name__: natural_log
+            natural_log.__name__: natural_log,
```
Comment (lastmansleeping) [inline python/ml4ir/base/features/preprocessing.py:21]: @mohazahran You should add `convert_label_to_clicks`. Why did you remove it?

File: python/ml4ir/base/pipeline.py (hunk 20 above)
Comment (lastmansleeping) [reply python/ml4ir/base/pipeline.py:274]: Since converting graded relevance to clicks is only a Ranking specific operation, it makes sense to keep it out of the `base/pipeline.py` and move it to `applications/ranking/pipeline.py`

Comment (lastmansleeping) [reply python/ml4ir/base/pipeline.py:274]: In fact, you don't have to do this explicitly. Let's quickly sync offline about this. ml4ir should automatically apply it if you add it to `PreProcessingMap` in `ml4ir/base/features/preprocessing.py`

**Review by lastmansleeping — APPROVED** [review]
Looks good. Thanks for incorporating the changes. I will merge and QA this. We can create follow up PR if there are issues.
